python/lsst/meas/extensions/scarlet/io.py

```diff
@@ -472,3 +482,153 @@ def oldScarletToData(blend: Blend, psfCenter: tuple[int, int], xy0: Point2I):
+            Dictionary mapping filenames to the desired file type.
+
+        Returns
+        -------
+        model :
+            ScarletModelData instance.
+        """
+        model_psf = zip_dict['psf.json']
+        psf_shape = zip_dict['psf_shape.json']
+        blends = {}
+        for key, value in zip_dict.items():
+            if key.startswith('blend_'):
+                blend_id = int(key.split('.')[0].split('_')[1])
+                blends[blend_id] = value
+        return scl.io.ScarletModelData.parse_obj({
+            'psf': model_psf,
+            'psfShape': psf_shape,
+            'blends': blends,
+        })
+
+    def _model_to_zip(self, model_data: scl.io.ScarletModelData) -> dict[str, Any]:
+        """Convert a ScarletModelData instance to a dictionary of files.
+
+        Parameters
+        ----------
+        model_data : `scl.io.ScarletModelData`
+            ScarletModelData instance.
+
+        Returns
+        -------
+        zip_dict :
+            Dictionary mapping filenames to the desired file type.
+        """
+        json_model = model_data.as_dict()
+
+        data = {
+            f'blend_{blend_id}.json': json.dumps(blend_data)
+            for blend_id, blend_data in json_model['blends'].items()
+        }
+        data.update({
+            'psf_shape.json': json.dumps(json_model['psfShape']),
+            'psf.json': json.dumps(json_model['psf']),
+        })
+        return data
+
+    def read_from_uri(self, uri: ResourcePath, component: str | None = None, expected_size: int = -1) -> Any:
+        # Override of `FormatterV2.read_from_uri`.
+        if self.file_descriptor.parameters is not None and "blend_id" in self.file_descriptor.parameters:
+            blend_id = self.file_descriptor.parameters["blend_id"]
+            blend_filename = f"blend_{blend_id}.json"
+        else:
+            blend_filename = None
+
+        if blend_filename is None:
+            unzipped_files = self._extract_all_files(uri)
+        else:
+            unzipped_files = self._extract_blend(uri, blend_filename)
+
+        return self._build_model(unzipped_files)
+
+    def to_bytes(self, in_memory_dataset: Any) -> bytes:
+        in_memory_zip = BytesIO()
+        with zipfile.ZipFile(in_memory_zip, 'w') as zf:
+            zip_archive = self._model_to_zip(in_memory_dataset)
+            for filename, data in zip_archive.items():
+                zf.writestr(filename, data)
+        return in_memory_zip.getvalue()
+
+
+class ScarletModelDelegate(StorageClassDelegate):
+    def can_accept(self, inMemoryDataset: Any) -> bool:
+        return isinstance(inMemoryDataset, scl.io.ScarletModelData)
+
+    def getComponent(self, composite: Any, componentName: str) -> Any:
+        raise AttributeError(f"Unsupported component: {componentName}")
+
+    def handleParameters(self, inMemoryDataset: Any, parameters: Mapping[str, Any] | None = None) -> Any:
+        return inMemoryDataset
```

Comment:
Shouldn't this be handling the blend_id parameter and returning a subset?

Reply:
I wasn't sure how this worked/if it was necessary. My thinking was that since fetching the data will have already handled the parameters to load only the selected blends, there is nothing that needs to be done here. Is that assumption wrong?

Reply:
Yes. Wrong. The delegate is what is used when a file datastore isn't involved. This can be an in-memory datastore or an InMemoryDatasetHandle (that is very useful for testing where people have written tasks that use a deferred dataset handle but don't want to use a butler). This method will be given the main object and then have to deal with the parameters to return the subset.

See for example the delegate for Exposure: https://github.com/lsst/obs_base/blob/main/python/lsst/obs/base/exposureAssembler.py#L280

Reply:
Ah, ok, thanks. Is there a way to test this in a unit test? I mean it's a trivial piece of code but it's still good to be able to test it.

Reply:
Yes. Use pipe base InMemoryDatasetHandle. (now commenting in the right place)